[Style] Convert the 'align-*'/'justify-*' properties to strong style types

weinig opened the pull request:
#### aefa156346394392a734e686c81f99d59da0ad05
<pre>
[Style] Convert the &apos;align-*&apos;/&apos;justify-*&apos; properties to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=302092#">https://bugs.webkit.org/show_bug.cgi?id=302092#</a>
Reviewed by Antti Koivisto and Darin Adler.

Converts the &apos;align-*&apos;/&apos;justify-*&apos; alignment properties to use strong style types.

As a lot of rendering code uses the existing alignment types, `StyleSelfAlignmentData`
and `StyleContentAlignmentData`, the new type are resolved to the older types for use.

This change also updates the parser to only use `CSSPrimitiveValue` and `CSSValuePair`
for the CSS specified representation. The old `CSSContentDistributionValue` made things
more complicated and didn&apos;t provide any concrete value. In addition, the parser was
updated to use switch statements more, making it more in line with other parts of the
parsing infrastructure. This introduces a bit of duplication, but overall makes each
property more clear.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSContentDistributionValue.cpp: Removed.
* Source/WebCore/css/CSSContentDistributionValue.h: Removed.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp:
* Source/WebCore/layout/formattingContexts/flex/FlexLayout.cpp:
* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp:
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderMenuList.cpp:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTableCellInlines.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleContentAlignmentData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleSelfAlignmentData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/align/StyleAlignContent.cpp: Added.
* Source/WebCore/style/values/align/StyleAlignContent.h: Added.
* Source/WebCore/style/values/align/StyleAlignItems.cpp: Added.
* Source/WebCore/style/values/align/StyleAlignItems.h: Added.
* Source/WebCore/style/values/align/StyleAlignSelf.cpp: Added.
* Source/WebCore/style/values/align/StyleAlignSelf.h: Added.
* Source/WebCore/style/values/align/StyleBaselineAlignmentPreference.h: Added.
* Source/WebCore/style/values/align/StyleJustifyContent.cpp: Added.
* Source/WebCore/style/values/align/StyleJustifyContent.h: Added.
* Source/WebCore/style/values/align/StyleJustifyItems.cpp: Added.
* Source/WebCore/style/values/align/StyleJustifyItems.h: Added.
* Source/WebCore/style/values/align/StyleJustifySelf.cpp: Added.
* Source/WebCore/style/values/align/StyleJustifySelf.h: Added.
* Source/WebCore/style/values/align/StyleLegacyPosition.h: Added.
* Source/WebCore/style/values/align/StyleOverflowPosition.h: Added.
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:

Canonical link: <a href="https://commits.webkit.org/302800@main">https://commits.webkit.org/302800@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/021ce824a09853dbf300e60a48500bea7aeeed55

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130131 "18 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2402 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41085 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137531 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81674 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132002 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2374 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2290 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99136 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66941 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133078 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1788 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116558 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79830 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1706 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34685 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80805 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110222 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35194 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140010 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2192 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2050 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107661 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2236 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112902 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107539 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27400 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1746 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31352 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55079 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2262 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65649 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2079 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2283 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2188 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -90,26 +90,36 @@ AnchorScrollAdjuster::AnchorScrollAdjuster(RenderBox& anchored, const RenderBoxM
+            }
         }
         if (positionArea->coordMatchedTrackForAxis(BoxAxis::Vertical, containingWritingMode, style.writingMode()) != PositionAreaTrack::SpanAll)
             m_needsYAdjustment |= true;
```
How is this (existing) code different from `= true`?

weinig replied:
I don't think it is.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp:
```diff
@@ -39,206 +38,455 @@
-{
-    return identMatches<CSSValueStart, CSSValueEnd, CSSValueCenter, CSSValueFlexStart, CSSValueFlexEnd>(id);
-}
+    // FIXME: The spec states that <baseline-position> is defined as `<baseline-position> = [ first | last ]? && baseline`, allowing any ordering, but tests expect `[ first | last ]` to always be precede `baseline`.
```
Are you going to file a spec issue about this?

weinig replied:
Good point.

weinig committed:
[Style] Convert the 'align-*'/'justify-*' properties to strong style types
https://bugs.webkit.org/show_bug.cgi?id=302092#
Reviewed by Antti Koivisto and Darin Adler.

Converts the 'align-*'/'justify-*' alignment properties to use strong style types.

As a lot of rendering code uses the existing alignment types, `StyleSelfAlignmentData`
and `StyleContentAlignmentData`, the new type are resolved to the older types for use.

This change also updates the parser to only use `CSSPrimitiveValue` and `CSSValuePair`
for the CSS specified representation. The old `CSSContentDistributionValue` made things
more complicated and didn't provide any concrete value. In addition, the parser was
updated to use switch statements more, making it more in line with other parts of the
parsing infrastructure. This introduces a bit of duplication, but overall makes each
property more clear.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSContentDistributionValue.cpp: Removed.
* Source/WebCore/css/CSSContentDistributionValue.h: Removed.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp:
* Source/WebCore/layout/formattingContexts/flex/FlexLayout.cpp:
* Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp:
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
* Source/WebCore/layout/integration/grid/LayoutIntegrationGridCoverage.cpp:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderMenuList.cpp:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTableCellInlines.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleContentAlignmentData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleSelfAlignmentData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/align/StyleAlignContent.cpp: Added.
* Source/WebCore/style/values/align/StyleAlignContent.h: Added.
* Source/WebCore/style/values/align/StyleAlignItems.cpp: Added.
* Source/WebCore/style/values/align/StyleAlignItems.h: Added.
* Source/WebCore/style/values/align/StyleAlignSelf.cpp: Added.
* Source/WebCore/style/values/align/StyleAlignSelf.h: Added.
* Source/WebCore/style/values/align/StyleBaselineAlignmentPreference.h: Added.
* Source/WebCore/style/values/align/StyleJustifyContent.cpp: Added.
* Source/WebCore/style/values/align/StyleJustifyContent.h: Added.
* Source/WebCore/style/values/align/StyleJustifyItems.cpp: Added.
* Source/WebCore/style/values/align/StyleJustifyItems.h: Added.
* Source/WebCore/style/values/align/StyleJustifySelf.cpp: Added.
* Source/WebCore/style/values/align/StyleJustifySelf.h: Added.
* Source/WebCore/style/values/align/StyleLegacyPosition.h: Added.
* Source/WebCore/style/values/align/StyleOverflowPosition.h: Added.
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:

Canonical link: https://commits.webkit.org/302800@main
